src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -635,11 +622,23 @@ public void Add(T item, EntryState state)
                 public TableEntry ToImmutableAndFree()
                 {
                     Debug.Assert(_currentState.HasValue, "Created a builder with no values?");
+                    Debug.Assert(_items.Count >= 1, "Created a builder with no values?");
 
                     Debug.Assert(_items.Count == _requestedCapacity);
                     Debug.Assert(_states == null || _states.Count == _requestedCapacity);
 
-                    return new TableEntry(item: default, _items.ToImmutableAndFree(), _states?.ToImmutableAndFree() ?? GetSingleArray(_currentState.Value));
+                    var states = _states?.ToImmutableAndFree() ?? GetSingleArray(_currentState.Value);
+
+                    if (_items.Count == 1)
+                    {
+                        var item = _items[0];
+                        _items.Free();
+                        return new TableEntry(OneOrMany.Create(item), states);
+                    }
+                    else
+                    {
+                        return new TableEntry(OneOrMany.Create(_items.ToImmutableAndFree()), states);
+                    }
```

Comment:
Minor: Consider only calculating `OneOrMany<T> items` in the `if`, with a single `return new TableEntry(items, states);` after.

Reply:
yup. can do in followup.